### PR TITLE
[Merged by Bors] - feat(FermatLastTheorem): Add a relaxed variant of FLT allowing nonzero unit solutions 

### PR DESCRIPTION
**Pull request by jcpaik**
Adds a variant `FermatLastTheoremWith'` of FLT for general commutative semirings, allowing nonzero unit solutions and their multiples.

---

Supersedes #14270. I find this version less intrusive to the original statement of FLT.

The relaxed 'unit' variant `FermatLastTheoremWith' α` states that any nonzero solution to Fermat equation $a^n+b^n=c^n$ should be units $(a, b, c)$ or their common multiples. The stronger original `FermatLastTheoremWith α` states that there is no nonzero solution $(a, b, c)$ from the beginning.

For `α = ℕ` or `ℤ` the two statements are equivalent to the original FLT. A purpose of this is to include a polynomial version of FLT which is easy to prove (relevant Zulip discussion [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Mason-Stothers.20theorem.20and.20Polynomial.20FLT)). `FermatLastTheoremWith k[X]` is false for general field $k$, but `FermatLastTheoremWith' k[X]` is.
 
I speculate that for some integral ring `α` over `ℤ`, the original statement of FLT is false while the suggested unit variant is true. For example, take the integral closure of $\mathbb{Z}[w]$ where $w = (-1 + \sqrt{3} i ) / 2$ is the 3rd root of unity satisfying $w^3 = 1$. The original statement for $n = 6k + 1$ is false as $w^n + (w^2)^n = (-1)^n$, but note that the solutions only consists of units.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 1]
```diff
@@ -153,6 +153,74 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
...
+3. For a polynomial ring `α = k[X]`, the original `FermatLastTheoremWith α` is false but the weaker
+   variant `FermatLastTheoremWith' α` is true. This polynomial variant of Fermat's Last Theorem
+   can be shown elementarily using Mason--Stothers theorem.
+-/
+def FermatLastTheoremWith' (α : Type*) [CommSemiring α] (n : ℕ) : Prop :=
+  ∀ a b c : α, a ≠ 0 → b ≠ 0 → c ≠ 0 → a ^ n + b ^ n = c ^ n →
+    ∃ d a' b' c', (a = a' * d ∧ b = b' * d ∧ c = c' * d) ∧ (IsUnit a' ∧ IsUnit b' ∧ IsUnit c')
+
+lemma fermatLastTheoremWith'_of_fermatLastTheoremWith {α : Type*} [CommSemiring α] {n : ℕ}
+    (h : FermatLastTheoremWith α n) : FermatLastTheoremWith' α n :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Basic.lean:173]: Probably this should be named to allow dot notation: `FermatLastTheoremWith.fermatLastTheoremWith'`

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 2]
```diff
@@ -153,6 +153,74 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
...
+  fun a b c _ _ _ _ ↦ by exfalso; apply h a b c <;> assumption
+
+lemma fermatLastTheoremWith'_of_field (α : Type*) [Field α] (n : ℕ) : FermatLastTheoremWith' α n :=
+  fun a b c ha hb hc _ ↦
+    ⟨1, a, b, c,
+     ⟨(mul_one a).symm, (mul_one b).symm, (mul_one c).symm⟩,
+     ⟨ha.isUnit, hb.isUnit, hc.isUnit⟩⟩
+
+lemma fermatLastTheoremWith_of_fermatLastTheoremWith' {α : Type*} [CommSemiring α] [IsDomain α]
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Basic.lean:182]: Same: `FermatLastTheoremWith'.fermatLastTheoremWith`

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 3]
```diff
@@ -153,6 +153,74 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
...
+    FermatLastTheoremWith α n := by
+  intro a b c ha hb hc heq
+  rcases h a b c ha hb hc heq with ⟨d, a', b', c', ⟨eq_a, eq_b, eq_c⟩, ⟨ua, ub, uc⟩⟩
+  rw [eq_a, eq_b, eq_c, mul_pow, mul_pow, mul_pow, ← add_mul, mul_eq_mul_right_iff] at heq
+  cases' heq with heq hd
+  · exact hn _ _ _ ua ub uc heq
+  · obtain ⟨hd, hn⟩ := pow_eq_zero_iff'.mp hd
+    rw [eq_a, hd, mul_zero] at ha
+    apply ha
+    rfl
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Basic.lean:194]: Slightly shorter:
```suggestion
    (hn : ∀ {a b c : α}, IsUnit a → IsUnit b → IsUnit c → a ^ n + b ^ n ≠ c ^ n) :
    FermatLastTheoremWith α n := by
  intro a b c ha hb hc heq
  rcases h a b c ha hb hc heq with ⟨d, a', b', c', ⟨rfl, rfl, rfl⟩, ⟨ua, ub, uc⟩⟩
  rw [mul_pow, mul_pow, mul_pow, ← add_mul] at heq
  exact hn ua ub uc <| mul_right_cancel₀ (pow_ne_zero _ (right_ne_zero_of_mul ha)) heq
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
Some small comments:

**Review by Ruben-VandeVelde — APPROVED** [review]
No objection from me.

maintainer merge

**Comment by kbuzzard** [comment]
Yeah this is all fine by me. To be quite honest I do think this work is interesting -- it makes formal the statement "Mason-Stothers is FLT for polynomial rings". The naive generalization is false (loads of constant solutions).

bors merge

**Pull request closed without merging**